---
title: "ADNP ChIP mapping and peak finding"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

# merge all Inputs to get deeply sequenced sample with decent coverage for IP/Input calculations

```{bash eval=FALSE}
cd /deepSeqData/bam

module load SAMtools/1.17-GCC-11.3.0

samtools merge -o Input_AGBTP_merged_RS_multi.bam \
Adnp_P3_In_AdnpWT_3676F13_RS_multi.bam  \
Gtf3c1FKBP_In_ctrl_r3_3760F10_RS_multi.bam Gtf3c1FKBP_In_ctrl_r4_3760F11_RS_multi.bam \
Gtf3c1FKBP_In_ctrl_r1_3578F9_RS_multi.bam Gtf3c1FKBP_In_ctrl_r2_3578F10_RS_multi.bam \
Adnp_P3_In_AdnpFKBP_3597F15_RS_multi.bam \
Gtf3c1fl_In_AdnpFKBP_ctrl_3730F21_RS_multi.bam Gtf3c1fl_In_AdnpFKBP_24h_3730F22_RS_multi.bam \
Gtf3afl_In_AdnpFKBPV5_ctrl_3848F25_RS_multi.bam Gtf3afl_In_AdnpFKBPV5_24h_3848F26_RS_multi.bam -@ 12

samtools index Input_AGBTP_merged_RS_multi.bam
```

# control ChIPs

```{bash eval=FALSE}
FLAG-ChIPs in untagged background:
wt_FLAGChIP_mm39_r1
wt_FLAGChIP_mm39_r2
wt_FLAGChIP_mm39_r3
```

# ADNP ChIP

```{r eval=FALSE}
Adnp_ChIP_AdnpWT_r1
Adnp_ChIP_AdnpWT_r2
Adnp_ChIP_AdnpWT_r3
Adnp_ChIP_AdnpWT_r4
Adnp_ChIP_AdnpWT_r5
Adnp_ChIP_AdnpWT_r6
```


```{bash eval=FALSE}
cd /michi/Projects/Adnp/Pol3_mm39/manuscript
bamdir=/deepSeqData/bam
module purge
module load MACS3/3.0.1
macs3 callpeak -t \
${bamdir}/Adnp_ChIP_AdnpWT_r1_3676F2_RS_multi.bam \
${bamdir}/Adnp_ChIP_AdnpWT_r2_3676F3_RS_multi.bam \
${bamdir}/Adnp_ChIP_AdnpWT_r3_3273F7_RS_multi.bam \
${bamdir}/Adnp_ChIP_AdnpWT_r4_3273F8_RS_multi.bam \
${bamdir}/Adnp_ChIP_AdnpWT_r5_3273F9_RS_multi.bam \
${bamdir}/Adnp_ChIP_AdnpWT_r6_3273F10_RS_multi.bam \
-f BAMPE -g mm --outdir peaks -n Adnp_WT

# FLAG ChIPs in untagged wt ESCs as background ctrl for FLAG antibody (peaks mostly over simple/low complexity repeats)
cd /michi/Projects/Adnp/Pol3_mm39/manuscript
bamdir=/deepSeqData/bam
module purge
module load MACS3/3.0.1
macs3 callpeak -t \
${bamdir}/wt_FLAGChIP_mm39_r1_3581F50_RS_multi.bam \
${bamdir}/wt_FLAGChIP_mm39_r2_3583F2_RS_multi.bam \
${bamdir}/wt_FLAGChIP_mm39_r3_3583F50_RS_multi.bam \
-f BAMPE -g mm --outdir peaks -n FLAGChIP_untagged
```



# calculate ChIP/Input enrichments in peaks and filter out peaks with low enrichments

```{r libraries}
suppressPackageStartupMessages({
  library(tidyverse)
  library(GenomicRanges)
  library(cowplot)
  library(QuasR)
  library(Rsubread)

})

#parameters
#topdir <-  "../"
topdir <- "/michi/Projects/Adnp/Pol3_mm39/manuscript"
datadir <- "data"
bamdir <- "/deepSeqData/bam"

genome_file <- "/bioinfo/Genomes/mm39/mm39.fa"
chrsizes_file <-"/bioinfo/Genomes/mm39/starIndex_v2_7_3a_mm39.woGTF/chrNameLength.txt"
repeat_file <- "/bioinfo/Annotations/mm39/repeat_masker_s/GRCm39.primary_assembly.repMasker_sesitive.bed"
gene_file <- "/bioinfo/Genomes/mm39/Gencode_GRCm39_M34/gencode.vM34.primary_assembly.annotation.gtf"
```

### load peaks, remove peaks overlapping FLAG ChIP artefacts and peaks >5kb and center them arround summit

```{r peaks, results=FALSE}
## peaks from all 6 replicates
peaks.df <- read.table(file.path(topdir,"peaks/Adnp_WT_peaks.narrowPeak"),header=F)
names(peaks.df) <- c("chr","start","end","name","score","empty","foldchange","pvalue","qvalue","summit")
peaks <- makeGRangesFromDataFrame(peaks.df,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("chr"),
                                   start.field=c("start"),
                                   end.field=c("end"),
                                   starts.in.df.are.0based=TRUE)


# peaks from FLAG ChIP in untagged wt ESCs
FLAGpeaks.df <- read.table(file.path(topdir,"peaks/FLAGChIP_untagged_peaks.narrowPeak"),header=F)
names(FLAGpeaks.df) <- c("chr","start","end","name","score","empty","foldchange","pvalue","qvalue","summit")
FLAGpeaks <- makeGRangesFromDataFrame(FLAGpeaks.df,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("chr"),
                                   start.field=c("start"),
                                   end.field=c("end"),
                                   starts.in.df.are.0based=TRUE)



# remove peaks overlapping FLAGonly peak
peaks <- subsetByOverlaps(peaks, FLAGpeaks, invert = TRUE)

#remove peaks > 5kb)
peaks <- peaks[width(peaks) < 5001]

#center peaks on summit
start(peaks) <- start(peaks) + peaks$summit
end(peaks) <- start(peaks)
peaks <- resize(peaks,width=300L, fix="center")
names(peaks) <- peaks$name

#remove peaks that extend over the edge of chromosome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
peaks <- peaks[overlapsAny(peaks,chr_sizes,minoverlap = 300)==TRUE]
common_levels <- seqlevels(peaks)[seqlevels(peaks) %in% seqlevels(chr_sizes)]
peaks <- keepSeqlevels(peaks,value = common_levels)
```

### calculate read counts in peaks for ChIPs and Inputs

```{r counts}
#get bam files
bamFiles <- list.files(bamdir,pattern = "_RS_multi.bam$",full.names = TRUE)
bamFiles <- c(grep("Adnp_ChIP_AdnpWT_r1_3676F2",bamFiles,value=TRUE),
              grep("Adnp_ChIP_AdnpWT_r2_3676F3",bamFiles,value=TRUE),
              grep("Adnp_ChIP_AdnpWT_r3_3273F7",bamFiles,value=TRUE),
              grep("Adnp_ChIP_AdnpWT_r4_3273F8",bamFiles,value=TRUE),
              grep("Adnp_ChIP_AdnpWT_r5_3273F9",bamFiles,value=TRUE),
              grep("Adnp_ChIP_AdnpWT_r6_3273F10",bamFiles,value=TRUE),
              grep("Input_AGBTP_merged",bamFiles,value=TRUE))
bamFiles

bamNames <- gsub("/deepSeqData/bam/","",bamFiles)
bamNames <- gsub("_RS_multi.bam","",bamNames)
bamNames

#saf table for featureCounts
saf <- data.frame(GeneID= names(peaks), Chr=seqnames(peaks),
                  Start=start(peaks), End=end(peaks),Strand=strand(peaks),Score=peaks$score)

#feature counts using only unique reads
f_counts <- featureCounts(bamFiles,annot.ext=saf,isGTFAnnotationFile = FALSE,
                          useMetaFeatures=FALSE,allowMultiOverlap=TRUE,
                          minOverlap=1,countMultiMappingReads=FALSE,fraction=TRUE,
                          minMQS=255,strandSpecific=0,nthreads=24,verbose=FALSE,isPairedEnd=TRUE)

fcounts <- data.frame(f_counts$counts)
colnames(fcounts) <- bamNames


```

### normalize counts to total mapped read numbers (cpm)

```{r cpmNormalisation}

#calculate CPMs
cpm <- f_counts$counts

#normalize to 1 mio mappers per library
columns = c("libs","mapped_reads") 
mapped_reads_df = data.frame(matrix(nrow = length(bamNames), ncol = length(columns))) 
colnames(mapped_reads_df) = columns
mapped_reads_df$libs <- bamNames

i=1
for (i in seq_along(bamFiles)){
  mapped_reads <- sum(f_counts$stat[f_counts$stat$Status=="Assigned" | f_counts$stat$Status=="Unassigned_NoFeatures",i+1])
  mapped_reads_df$mapped_reads[i] <- mapped_reads
  scaling_factor <-  mapped_reads/1e6
  cpm[,i] <- f_counts$counts[,i]/scaling_factor
  i = i+1
}


cpm <- data.frame(cpm)
colnames(cpm) <- bamNames


```

### calculate ChIP/Input enrichment

```{r enrichments}
#| fig-width: 10
#| fig-height: 2

Adnp_cpmsum <- (cpm[,1:6])+(cpm[,7])
Adnp_enrich <- log2((cpm[,1:6]+1)/(cpm[,7]+1))

# plot enrichement vs peak score from MACS3
smoothScatter(rowMeans(Adnp_enrich[,1:6]),log2(peaks$score))
abline(h=log2(50),col="gold",lty=2)
abline(h=log2(100),col="darkorange2",lty=2)
abline(h=log2(200),col="red",lty=2)


# MA plots 
ids <- colnames(Adnp_enrich)
ylims <- range(Adnp_enrich)
xlims <- range(log2(Adnp_cpmsum+1))

pL <- lapply(ids[seq(1, length(ids))], function(id) {
  ggplot(data.frame(M=Adnp_enrich[,id],A=log2(Adnp_cpmsum[,id]+1)),aes(A,M)) + geom_point() + ggtitle(id) + theme_bw() + xlim(xlims) + ylim(ylims) + geom_hline(yintercept=0,col="firebrick",linetype="dashed")
})
plot_grid(plotlist = pL, ncol = 4, nrow = 2, align = "hv")


# density plots of enrichments
ids <- colnames(Adnp_enrich)
xlims <- range(Adnp_enrich)

pL <- lapply(ids[seq(1, length(ids))], function(id) {
  ggplot(data.frame(M=Adnp_enrich[,id],A=log2(Adnp_cpmsum[,id]+1)),aes(M)) + geom_density() + ggtitle(id) + theme_bw() + geom_vline(xintercept=log2(1.2),col="firebrick",linetype="dashed") + xlim(xlims)
})
plot_grid(plotlist = pL, ncol = 4, nrow = 2, align = "hv")
```


### filter peaks based on enrichment (more than 1.2 fold enriched in at least 3 out of 6 replicates)


```{r filter}
# filter cpm and fe (finally used!)
Adnp_cpm <- Adnp_cpmsum > 0.8
Adnp_enr <- Adnp_enrich > log2(1.2)
Adnp_Peaks_filtered <- peaks[rowSums(Adnp_enr) >=3 & rowSums(Adnp_cpm) >=3]
Adnp_enr_filtered <- Adnp_enrich[rowSums(Adnp_enr) >=3 & rowSums(Adnp_cpm) >=3,]

```


# compare MACS3 score before and after filtering, and remove peaks with MACS peak score < 100

```{r}

Adnp_Peaks_filtered_final <- Adnp_Peaks_filtered[Adnp_Peaks_filtered$score > 100]
Adnp_enr_filtered_final <- Adnp_enr_filtered[Adnp_Peaks_filtered$score > 100,]

plot(density(log2(peaks$score)), ylim=c(0,0.4), xlim=c(0,15), main="density plot of MACS peak score before and after filtering")
lines(density(log2(Adnp_Peaks_filtered$score)),col="red")
abline(v=log2(100),col="darkgreen",lty=2)

```



# save the peaks

```{r}
# save final peak set: 1.2 foldEnr, >05 cpm, score >100
mcols(Adnp_Peaks_filtered_final) <- cbind(mcols(Adnp_Peaks_filtered_final),Adnp_enr_filtered_final)
saveRDS(object = Adnp_Peaks_filtered_final, file = file.path(topdir,datadir, "Adnp_Peaks_filtered_1.2foldEnr_05cpm_score100.rds"))

#save as bed file
Adnp_Peaks_filtered_bed <- data.frame(chr=seqnames(Adnp_Peaks_filtered_final),start=start(Adnp_Peaks_filtered_final)-1,end=end(Adnp_Peaks_filtered_final),
           names=Adnp_Peaks_filtered_final$name,
                 scores=Adnp_Peaks_filtered_final$score,
                 strands=strand(Adnp_Peaks_filtered_final))
write.table(Adnp_Peaks_filtered_bed, file=file.path(topdir,"peaks/Adnp_Peaks_filtered_1.2foldEnr_05cpm_score100.bed"), quote=F, sep="\t", row.names=F, col.names=F)


# compare old and new peaks (qCount vs featurecount filtering)
require(ChIPpeakAnno)
old_peaks <- readRDS(file = file.path(topdir,datadir, "Adnp_Peaks_filtered_1.2foldEnr_05cpm_score100.rds"))
makeVennDiagram(Peaks=list(Adnp_Peaks_filtered_final,old_peaks), NameOfPeaks = c("new","old") ,totalTest = 100000)

```

# intersect peaks with repeat and gene annotations
```{r}
# libs
require(TxDb.Mmusculus.GENCODE.GRCm39.gencodeM34)
require(rtracklayer)

#load peaks
peaks <- readRDS(file = file.path(topdir,datadir, "Adnp_Peaks_filtered_1.2foldEnr_05cpm_score100.rds"))

# get gene symbols for Genoced M34
GFF <- readGFF("/bioinfo/Genomes/mm39/Gencode_GRCm39_M34/gencode.vM34.primary_assembly.annotation.gtf")
gene_names <- unique(GFF[,c(9:11)])
# add as metacolumns
genes <- genes(TxDb.Mmusculus.GENCODE.GRCm39.gencodeM34)
mcols(genes) <- merge(mcols(genes),gene_names)
# check if order was maintained
identical(names(genes),genes$gene_id)

# get out protein coding genes and non-coding RNAs
genesProt <- genes[genes$gene_type=="protein_coding"]
ncRNAs <- genes[ grepl("RNA",genes$gene_type)]
genebodies <- genesProt[width(genesProt)>200]
start(genebodies) <- start(genebodies)+200

# get out TSSs
TSSgenes <- promoters(genesProt,upstream=800, downstream=200)
TSSnc <- promoters(ncRNAs,upstream=800, downstream=200)
TSSs <- c(TSSgenes,TSSnc[!overlapsAny(TSSnc,TSSgenes)])

# add metacolumns to merge with rep masker annotations
TSSs$repeat_class <- "TSS"
TSSs$repeat_family <- "TSS"
TSSs$repeat_classfam <- "TSS/TSS"
names(mcols(TSSs))[9] <- "repeat_class/family"
TSSs$percDiv <- 100
TSSs$percDel <- 1000
TSSs$percIns <- 1000


#load repeat annotations from repeat masker
repeats <- read.delim(repeat_file,header=FALSE)
names(repeats) <- c("chr","start","end","repeat_name","swScore","strand","repeat_class","repeat_family","percDiv","percDel","percIns","repeat_class/family")
reps <- makeGRangesFromDataFrame(repeats,
                                 keep.extra.columns=TRUE,
                                 ignore.strand=FALSE,
                                 seqinfo=NULL,
                                 seqnames.field=c("chr"),
                                 start.field=c("start"),
                                 end.field=c("end"),
                                 strand.field=c("strand"),
                                 starts.in.df.are.0based=TRUE)


#combine annotations
annot <- c(reps,TSSs)

#remove chromosomes not in combined genome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
annot <- annot[overlapsAny(annot,chr_sizes,minoverlap = 1)==TRUE]


#select the repeats that overlap peaks
peaks2reps <- findOverlaps(resize(peaks,width=50,fix="center"),annot,ignore.strand=TRUE)
peaks.reps <- peaks[from(peaks2reps)]
reps.peaks <- annot[to(peaks2reps)]
reps.peaks$peakID <- names(peaks.reps)

# remove the repeats that overlap the same peak, keep the onles that have a low percDiv and percIns and percDel
dup.peakIDs <- unique(reps.peaks$peakID[duplicated(reps.peaks$peakID)])

reps.peaks$score <- reps.peaks$percDiv*2 + reps.peaks$percDel + reps.peaks$percIns
uni.reps.peaks <- reps.peaks[reps.peaks$peakID %in% dup.peakIDs ==FALSE]

dup.reps.peaks <- lapply(dup.peakIDs,function(dup.peakIDs){
  sel.reps <- reps.peaks[reps.peaks$peakID==dup.peakIDs]
  sel.reps[sel.reps$score < (min(sel.reps$score)+0.01)]
#  sel.reps[sel.reps$score < (min(sel.reps$score)+5)]
})


#keep only one match, drop others as it is impossible to distinguish anyways
dup.reps.peaksGR <- do.call("c",dup.reps.peaks)
dups <- dup.reps.peaksGR$peakID[duplicated(dup.reps.peaksGR$peakID)]
dup.reps.peaksGR.uni <- dup.reps.peaksGR[duplicated(dup.reps.peaksGR$peakID)==FALSE]
uni.reps.peaks <- c(uni.reps.peaks,dup.reps.peaksGR.uni)
summary(as.factor(duplicated(uni.reps.peaks$peakID)))

#add the repeat name, family, class, divergence to peaks GRanges
peaks$peakID <- names(peaks)
peaks.df <- left_join(data.frame(peaks),data.frame(mcols(uni.reps.peaks)[,c(1,3,4,5,12)]),by="peakID")
peaks_wth_reps <- makeGRangesFromDataFrame(peaks.df[,-4],
                                 keep.extra.columns=TRUE,
                                 ignore.strand=FALSE,
                                 seqinfo=NULL,
                                 seqnames.field=c("seqnames"),
                                 start.field=c("start"),
                                 end.field=c("end"),
                                 strand.field = c("strand"),
                                 starts.in.df.are.0based=FALSE)
peaks_wth_reps$repclass <- ifelse(is.na(peaks_wth_reps$repeat_class),"non_repeat",
                              ifelse(grepl("^DNA",peaks_wth_reps$repeat_class),"DNA",
                                ifelse(grepl("^TSS",peaks_wth_reps$repeat_class),"TSS",
                                  ifelse(grepl("RNA",peaks_wth_reps$repeat_class) & peaks_wth_reps$repeat_class !="tRNA","ncRNA",
                                     ifelse(grepl("^tRNA",peaks_wth_reps$repeat_class),"tRNA",
                                        ifelse(grepl("^LINE",peaks_wth_reps$repeat_class),"LINE",
                                           ifelse(grepl("^LTR",peaks_wth_reps$repeat_class),"LTR",
                                                ifelse(grepl("^SINE",peaks_wth_reps$repeat_class),"SINE",
                                                       ifelse(grepl("Satellite",peaks_wth_reps$repeat_class),"Satellite",
                                                              ifelse(peaks_wth_reps$repeat_class=="Simple_repeat" | peaks_wth_reps$repeat_class=="Low_complexity","Simple/low_complexity","other"))))))))))

#check if there is no duplicated peaks
peaks_wth_reps[duplicated(peaks_wth_reps$peakID)]

summary(as.factor(peaks_wth_reps$repclass))

saveRDS(object = peaks_wth_reps, file = file.path(topdir,datadir, "ADNP_Peaks_filtered_1.2foldEnr_mq255_repAnnotated.rds"))

#peaks_wth_reps <- readRDS(file = file.path(topdir,datadir, "ADNP_Peaks_filtered_1.2foldEnr_mq255_repAnnotated.rds"))

```


# check some stats for reporting summary

```{r}
Adnp_Peaks_filtered <- readRDS(file = file.path(topdir,datadir, "Adnp_Peaks_filtered_1.2foldEnr_05cpm_score100.rds"))
Adnp_Peaks_filtered[Adnp_Peaks_filtered$qvalue > -log10(0.05)]
Adnp_Peaks_filtered[Adnp_Peaks_filtered$foldchange > 5]

```


# Session info

```{r session_info}
sessionInfo()
```

