---
title: "lonely SINEs"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}
suppressPackageStartupMessages({
  library(tidyverse)
  library(GenomicFeatures)
    library(Biostrings)
})

#parameters
topdir <-  "/michi/Projects/Adnp/Pol3_mm39/manuscript"
datadir <- "data"
plotdir <- "plots"
bamdir <- "/deepSeqData/bam"

genome_file <- "/bioinfo/Genomes/mm39/mm39.fa"
chrsizes_file <-"/bioinfo/Genomes/mm39/starIndex_v2_7_3a_mm39.woGTF/chrNameLength.txt"
options(ucscChromosomeNames=FALSE)
repeat_file <- "/bioinfo/Annotations/mm39/repeat_masker_s/GRCm39.primary_assembly.repMasker_sesitive.bed"

#make genome for mm39
mm39 <- readDNAStringSet(genome_file, format="fasta",
                 nrec=-1L, skip=0L, seek.first.rec=FALSE, use.names=TRUE)
names(mm39) <- matrix(unlist(strsplit(names(mm39)," ")),ncol=2,byrow=TRUE)[,1]
```

#load repeats

```{r}
#prepare repeat annotations from repeat masker run
repeats <- read.delim(repeat_file,header=FALSE)
names(repeats) <- c("chr","start","end","repeat_name","swScore","strand","repeat_class","repeat_family","percDiv","percDel","percIns","repeat_class/family")
reps <- makeGRangesFromDataFrame(repeats,
                                 keep.extra.columns=TRUE,
                                 ignore.strand=FALSE,
                                 seqinfo=NULL,
                                 seqnames.field=c("chr"),
                                 start.field=c("start"),
                                 end.field=c("end"),
                                 strand.field=c("strand"),
                                 starts.in.df.are.0based=TRUE)

#remove chromosomes not in combined genome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
reps <- reps[overlapsAny(reps,chr_sizes,minoverlap = 1)==TRUE]

names(reps) <- paste(reps$repeat_name,seqnames(reps),start(reps),end(reps),sep="_")

# keep only proper repeats
reps2 <- reps[reps$repeat_class %in% c("Simple_repeat","Unknown","Low_complexity","DNA?","ARTEFACT")==FALSE]

#remove duplicated repeats
reps2 <- reps2[!duplicated(names(reps2))]
reps2$ID <- names(reps2)

#POL3 transcribed repeats only
repsP3 <- reps2[reps2$repeat_class=="SINE" | reps2$repeat_class=="tRNA"]
```

# determine which repeats are without neighboring SINE or RNA repeats

```{r}
#200bp upstream window
reps2up <- promoters(reps2,upstream=200,downstream=0)
reps2up <- promoters(reps2up,upstream=5,downstream=195)

#200bp downstream window
reps2down <- terminators(reps2,upstream=0,downstream=200)
reps2down <- terminators(reps2down,upstream=195,downstream=5)

#select reps that have no Pol3 transcribed rep up or downstream
reps2$upstream <- overlapsAny(reps2up,repsP3,ignore.strand=TRUE)
reps2$downstream <- overlapsAny(reps2down,repsP3,ignore.strand=TRUE)

reps3 <- reps2[reps2$upstream==FALSE & reps2$downstream==FALSE]
write_rds(reps3,file.path(topdir,datadir,"lonely_repeats_200bp_noSINEstRNAs.rds"))
reps3 <- read_rds(file.path(topdir,datadir,"lonely_repeats_200bp_noSINEstRNAs.rds"))

```

#add A box, B box and CTCF motifs to SINEs 

#select SINEs

```{r}
SINEs <- reps3[reps3$repeat_class == "SINE"]
```


```{r}
library(motifmatchr)
library(TFBSTools)
load(file.path("/michi/Projects/Adnp/Pol3_mm39/",datadir,"A_and_B_box_PWMs.RData"))
SINEsseq <- BSgenome::getSeq(mm39,SINEs)

### A BOX
Abox_SINEs <- matchMotifs(Abox_PWM,SINEsseq,out="positions",p.cutoff = 5e-04)
#convert to GRanges
Abox_SINEs <- Abox_SINEs[[1]]
names(Abox_SINEs) <- names(SINEs)
Abox_SINE_GR <- as(Abox_SINEs,"GRanges")

#add strand and score column
Abox_SINE_unlist <- lapply(Abox_SINEs,mcols)
Abox_SINE_unlist <- do.call("rbind",Abox_SINE_unlist)
strand(Abox_SINE_GR) <- Abox_SINE_unlist$strand
Abox_SINE_GR$score <- Abox_SINE_unlist[,"score"]
Abox_SINE_DF <- unique(data.frame(ID=seqnames(Abox_SINE_GR),Abox_score=Abox_SINE_GR$score))
#remove duplicated IDs
Abox_SINE_DF <- Abox_SINE_DF %>% group_by(ID) %>% summarize(Abox_score=max(Abox_score))

#generate a data frame with the same rows as the SINEs Granges
SINE_motifs <- left_join(data.frame(ID=names(SINEs)),Abox_SINE_DF, by="ID")
#replace NAs
SINE_motifs <- replace_na(SINE_motifs,list(Abox_score=0))

### B BOX
Bbox_SINEs <- matchMotifs(Bbox_PWM,SINEsseq,out="positions",p.cutoff = 5e-04)
#convert to GRanges
Bbox_SINEs <- Bbox_SINEs[[1]]
names(Bbox_SINEs) <- names(SINEs)
Bbox_SINE_GR <- as(Bbox_SINEs,"GRanges")

#add strand and score column
Bbox_SINE_unlist <- lapply(Bbox_SINEs,mcols)
Bbox_SINE_unlist <- do.call("rbind",Bbox_SINE_unlist)
strand(Bbox_SINE_GR) <- Bbox_SINE_unlist$strand
Bbox_SINE_GR$score <- Bbox_SINE_unlist[,"score"]
Bbox_SINE_DF <- unique(data.frame(ID=seqnames(Bbox_SINE_GR),Bbox_score=Bbox_SINE_GR$score))
#remove duplicated IDs
Bbox_SINE_DF <- Bbox_SINE_DF %>% group_by(ID) %>% summarize(Bbox_score=max(Bbox_score))

#generate a data frame with the same rows as the SINEs Granges
SINE_motifs <- left_join(SINE_motifs,Bbox_SINE_DF, by="ID")
#replace NAs
SINE_motifs <- replace_na(SINE_motifs,list(Bbox_score=0))

## CTCF
library(JASPAR2022)
#CTCF
opts <- list()
opts[["name"]] <- "CTCF"
opts[["all_versions"]] <- TRUE
PFMatrixList <- getMatrixSet(JASPAR2022, opts)
PFMatrixList[[1]]
CTCF_PWM <- toPWM(PFMatrixList[[1]], pseudocounts=0.8)

CTCF_SINEs <- matchMotifs(CTCF_PWM,SINEsseq,out="positions",p.cutoff = 5e-04)
#convert to GRanges
CTCF_SINEs <- CTCF_SINEs[[1]]
names(CTCF_SINEs) <- names(SINEs)
CTCF_SINE_GR <- as(CTCF_SINEs,"GRanges")

#add strand and score column
CTCF_SINE_unlist <- lapply(CTCF_SINEs,mcols)
CTCF_SINE_unlist <- do.call("rbind",CTCF_SINE_unlist)
strand(CTCF_SINE_GR) <- CTCF_SINE_unlist$strand
CTCF_SINE_GR$score <- CTCF_SINE_unlist[,"score"]
CTCF_SINE_DF <- unique(data.frame(ID=seqnames(CTCF_SINE_GR),CTCF_score=CTCF_SINE_GR$score))
#remove duplicated IDs
CTCF_SINE_DF <- CTCF_SINE_DF %>% group_by(ID) %>% summarize(CTCF_score=max(CTCF_score))

#generate a data frame with the same rows as the SINEs Granges
SINE_motifs <- left_join(SINE_motifs,CTCF_SINE_DF, by="ID")
#replace NAs
SINE_motifs <- replace_na(SINE_motifs,list(CTCF_score=0))

#add to SINEs GRANGES
mcols(SINEs) <- cbind(mcols(SINEs),SINE_motifs)

#save SINEs GRanges
write_rds(SINEs,file.path(topdir,datadir,"lonely_SINEs_200bp_noSINEstRNAs_withABCmotifscores.rds"))

```

